BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,390 @@
+        at != cam_state.at ||
+        up != cam_state.up)
+    {
+        m_camera->LookAt(cam_state.pos, cam_state.at, cam_state.up);
+    }
+}
+
+void Render::SaveOutput(OutputInfo info,
+                        const std::string& name,
+                        bool gamma_correction_enabled,
+                        const std::filesystem::path& output_dir)
+{
+    OIIO_NAMESPACE_USING;
+
+    auto output = m_renderer->GetOutput(info.type);
+
+    assert(output);
+
+    auto buffer = static_cast<Baikal::ClwOutput*>(output)->data();
+
+    std::vector<RadeonRays::float3> output_data(buffer.GetElementCount());
+
+    output->GetData(output_data.data());
+
+    auto width = output->width();
+    auto height = output->height();
+
+    std::vector<float> image_data(info.chanels_num * width * height);
+
+    float* dst_row = image_data.data();
+
+    if (gamma_correction_enabled &&
+       (info.type == Renderer::OutputType::kColor) &&
+       (info.chanels_num == 3))
+    {
+        for (auto y = 0u; y < height; ++y)
+        {
+            for (auto x = 0u; x < width; ++x)
+            {
+                float3 val = output_data[(height - 1 - y) * width + x];
+                // "The 4-th pixel component is a count of accumulated samples.
+                //It can be different for every pixel in case of adaptive sampling.
+                //So, we need to normalize pixel values here".
+                val *= (1.f / val.w);
+                // gamma corection
+                dst_row[3 * x] = std::pow(val.x, 1.f / 2.2f);
+                dst_row[3 * x + 1] = std::pow(val.y, 1.f / 2.2f);
+                dst_row[3 * x + 2] = std::pow(val.z, 1.f / 2.2f);
+            }
+            dst_row += info.chanels_num * width;
+        }
+    }
+    else
+    {
+        for (auto y = 0u; y < height; ++y)
+        {
+            for (auto x = 0u; x < width; ++x)
+            {
+                float3 val = output_data[(height - 1 - y) * width + x];
+                // "The 4-th pixel component is a count of accumulated samples.
+                //It can be different for every pixel in case of adaptive sampling.
+                //So, we need to normalize pixel values here".
+                val *= (1.f / val.w);
+
+                if (info.chanels_num == 3)
+                {
+                    int dst_pixel = y * width + x;
+                    // invert the image 
+                    dst_row[3 * x] = val.x;
+                    dst_row[3 * x + 1] = val.y;
+                    dst_row[3 * x + 2] = val.z;
+                }
+                else // info.chanels_num = 1
+                {
+                    // invert the image 
+                    dst_row[x] = val.x;
+                }
+            }
+            dst_row += info.chanels_num * width;
+        }
+    }
+
+    std::filesystem::path file_name = output_dir;
+    file_name.append(name);
+
+    std::ofstream f;
+
+    f.open(file_name.string(), std::ofstream::binary);
```

Comment:
`std::ofstream f(file_name.string(), ...`
`f.close()` is not required.

Reply:
Done